File: sakinah-native/App/Sakinah/ScreenshotMode.swift
import Foundation
import SakinahCore

/// وضع لقطات المتجر: يُشغَّل بوسيط إقلاع وحده، فلا أثر له في نسخة المستخدم.
///
/// لقطات App Store **يجب أن تكون من التطبيق نفسه** (‏Guideline 2.3.3)، ولا تكفي
/// صورٌ من نسخة الويب لأن واجهتها ليست هذه الواجهة. وقيادة الواجهة باختبار
/// XCUITest تعني كتابة مُحدِّدات لنصوصٍ عربية تتغيّر — فالأضبط أن يفتح التطبيق
/// نفسه الشاشةَ المطلوبة مباشرةً، والوسيط هو الطريق المعتاد لذلك في iOS.
///
/// يُقرأ الوسيط بصيغة `-sakinahShot <route>` لأن UserDefaults تلتقط أزواج
/// `-مفتاح قيمة` من وسائط الإقلاع تلقائياً.
enum ScreenshotMode {
  /// المسار المطلوب، أو nil في التشغيل العادي
  static let route: String? = {
    guard let r = UserDefaults.standard.string(forKey: "sakinahShot"), !r.isEmpty else { return nil }
    return r
  }()
  static var active: Bool { route != nil }

  /// التبويب الذي يُفتح عليه التطبيق
  static var tab: AppTab? {
    switch route {
    case "prayer", "qibla", "home", "home-bottom", "sky-settings": return .home
    case "mushaf", "mushaf-juz", "mushaf-page", "mushaf-bar", "mushaf-ayah", "mushaf-nav", "mushaf-khatmah", "mushaf-display", "mushaf-hifz": return .mushaf
    case "adhkar", "hisn", "tasbih": return .adhkar
    case "hadith", "more": return .more
    default: return nil
    }
  }

  /// الصفحة التي يُفتح عليها قارئ المصحف، إن كان المسار يطلب القارئ
  static var readerPage: Int? { ["mushaf-page", "mushaf-bar", "mushaf-ayah", "mushaf-nav", "mushaf-display", "mushaf-hifz"].contains(route ?? "") ? 270 : nil }

  /// يُبقى الشريط ظاهرًا: القارئ يخفيه بعد ثوانٍ، فلا يلتقطه انتظارُ تحميل الخطوط
  static var keepChrome: Bool { route == "mushaf-bar" || route == "mushaf-ayah" }
  /// لقطات المصحف v2: آية محدّدة مع رصيفها، ورقة مفتوحة في القارئ، جلسة إخفاء للحفظ، وورقة الختمة في المكتبة
  static var readerSelectsAyah: Bool { route == "mushaf-ayah" }
  static var readerSheet: ReaderSheet? { switch route { case "mushaf-nav": return .quickNav; case "mushaf-display": return .display; default: return nil } }
  static var readerHifz: Bool { route == "mushaf-hifz" }
  static var librarySheet: Bool { route == "mushaf-khatmah" }
  /// تبويب الفهرس الذي تُفتح عليه المكتبة (الأجزاء في لقطة التحقّق من تبديل التبويبات)
  static var libraryTab: Int { route == "mushaf-juz" ? 1 : 0 }
  /// مسار «qibla» يفتح القبلة الكاملة فوق الرئيسية
  static var fullQibla: Bool { route == "qibla" }
  /// مسار «sky-settings» (تشخيصي): يفتح شاشة «مظهر السماء» فوق الرئيسية للتحقّق البصري منها
  static var skySettings: Bool { route == "sky-settings" }
  /// مسار «home-bottom» (تشخيصي لا للمتجر): يمرّر الرئيسية إلى آخرها كي تُرى آخر بطاقة فوق الشريط العائم
  /// — لقطة أعلى الصفحة لا تكشف تغطية الشريط لآخر بطاقة (قِيس في ثلاثة بناءات على جهاز المالك)
  static var scrollToBottom: Bool { route == "home-bottom" }
  /// لقطات التحقّق من نظام السماء: فرض طور (`-sakinahSky dhuhr`) وطقس (`-sakinahWeather rain`) — لا أثر لهما في التشغيل العادي
  static let skyPhase: SkyPhase? = UserDefaults.standard.string(forKey: "sakinahSky").flatMap { SkyPhase(rawValue: $0) }
  static let skyWeather: SkyWeather? = UserDefaults.standard.string(forKey: "sakinahWeather").flatMap { SkyWeather(rawValue: $0) }

  /// موقع ثابت كي تُحسب المواقيت والقبلة بلا إذنٍ ولا شبكة — واللقطات تتكرّر بالنتيجة نفسها
  static func seed(_ model: AppModel) {
    guard active else { return }
    model.settings.seenIntro = true
    // أرقام عربية-هندية: إعدادٌ يختاره المستخدم فعلاً، وصفحة المتجر عربية،
    // وصفحة المصحف نفسها ترقّم بها — فالخلط بين ٢٧٠ في الصفحة و270 في الشريط
    // يظهر في اللقطة تنافرًا لا داعي له.
    model.settings.numerals = "arab"
    if let c = CityDatabase.bundled.city(id: "jo-amman") { model.location.useCity(c) }
    // بطاقة «أقرب مسجد» مفعّلة: سير اللقطات يمنح المحاكي إذن الموقع وموقعًا محاكى (simctl privacy/location)
    // فتبحث البطاقة حول قراءة الجهاز كما عند المستخدم، وبلا ذلك تعرض طلب الإذن — وكلاهما يُتحقّق منه
    model.settings.nearbyMosques = true
    // ورقة الختمة والمكتبة تحتاجان خطةً وسجلّ ورد كي تُظهرا حالةً حقيقية: خطة ٣٠ يومًا بدأت قبل خمسة أيام وورد ٢١ صفحة في أربعة منها
    if route == "mushaf-khatmah" || route == "mushaf" {
      let today = model.todayKey
      model.quran.khatmah = KhatmahPlan(startPage: 1, startedAt: DayKey.adding(today, days: -5), days: 30, reminder: "after:isha")
      var log: WirdLog = [:]
      for (i, d) in [-5, -4, -3, -1, 0].enumerated() { for p in (i * 21 + 1)...((i + 1) * 21) { log = Wird.mark(log, today: DayKey.adding(today, days: d), page: p, startPage: 1) } }
      model.quran.wird = log
      model.settings.lastRead = LastRead(page: 106, surah: 5, ayah: 1, at: Date().timeIntervalSince1970 * 1000)
      model.quran.pushRecent(QuranText.shared.ayah(surah: 5, ayah: 1)!); model.quran.pushRecent(QuranText.shared.ayah(surah: 18, ayah: 10)!)
    }
  }
}

File: sakinah-android/app/src/main/kotlin/org/emdatra/sakinah/app/ScreenshotMode.kt
package org.emdatra.sakinah.app

import org.emdatra.sakinah.app.ui.AppTab
import org.emdatra.sakinah.core.CityDatabase

/**
 * وضع لقطات المتجر (Google Play) — يُفعَّل بمعطى الإطلاق فقط، ولا أثر له في الاستعمال العادي:
 *   adb shell am start -n org.emdatra.sakinah/.app.MainActivity --es sakinahShot mushaf-bar
 * المسارات نفسها التي في نسخة iOS كي تتطابق اللقطات بين المتجرين.
 */
object ScreenshotMode {
  var route: String? = null; private set
  val active: Boolean get() = route != null
  /** لقطات التحقّق من نظام السماء: فرض طور (`--es sakinahSky dhuhr`) وطقس (`--es sakinahWeather rain`) — لا أثر لهما في التشغيل العادي */
  var skyPhase: org.emdatra.sakinah.core.SkyPhase? = null; private set
  var skyWeather: org.emdatra.sakinah.core.SkyWeather? = null; private set
  val tab: AppTab? get() = when (route) {
    "prayer", "qibla", "home" -> AppTab.Home
    "mushaf", "mushaf-page", "mushaf-bar" -> AppTab.Mushaf
    "adhkar", "hisn", "tasbih" -> AppTab.Adhkar
    "hadith", "more", "sky-settings" -> AppTab.More
    else -> null
  }
  /** صفحة المصحف التي يُفتح عليها القارئ مباشرةً (سورة الشعراء — صفحة كثيفة الرسم) */
  val readerPage: Int? get() = if (route == "mushaf-page" || route == "mushaf-bar") 270 else null
  /** يُبقي الشريط الموحّد ظاهراً بدل انزلاقه بعد ثوانٍ */
  val keepChrome: Boolean get() = route == "mushaf-bar"
  /** مسار «qibla» يفتح القبلة الكاملة فوق الرئيسية */
  val fullQibla: Boolean get() = route == "qibla"

  /** تهيئة حالةٍ ثابتة: بلا مقدّمة، أرقام عربية، مدينة عمّان — فالمحاكي بلا موقع */
  fun arm(r: String?, sky: String? = null, weather: String? = null) {
    route = r?.takeIf { it.isNotEmpty() }
    if (!active) return
    skyPhase = sky?.let { id -> org.emdatra.sakinah.core.SkyPhase.entries.firstOrNull { it.id == id } }
    skyWeather = org.emdatra.sakinah.core.SkyWeather.of(weather)
    Store.seenIntro = true; Store.numerals = "arab"; Store.seenChromeHint = true
    CityDatabase.bundled.city("jo-amman")?.let { Store.useCity(it) }
    Store.save()
  }
}
